[JIT] Serialize attributes and types in ClassType serialization

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32658 [JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API
* #32604 [JIT] pickle serialization for custom bound classes
* #32586 [JIT] Include custom_class.h in torch/script.h
* **#32555 [JIT] Serialize attributes and types in ClassType serialization**
* #32553 [JIT] Support for registering C++ lambdas as methods on custom C++ class

This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

Commit:
[JIT] Serialize attributes and types in ClassType serialization

[ghstack-poisoned]

jamesr66a commented:
@suo 

Script
```
import torch

@torch.jit.script
class Foo:
    def __init__(self):
        self.x = 3


@torch.jit.script
def foo():
    f = Foo()
    return f.x

foo.save('test.pt')
```

Before
```
class Foo:
  def __init__(self: __torch__.Foo) -> None:
    self.x = 3
    return None
```

After
```
class Foo:
  x : int
  def __init__(self: __torch__.Foo) -> None:
    self.x = 3
    return None
```

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -270,6 +270,9 @@ void ivalue::Object::unsafeRemoveAttr(const std::string& name) {
 
 void ivalue::Object::resizeObject(size_t slot) {
+  if (slot >= type()->numAttributes()) {
+    std::cout << slot << " < " << type()->numAttributes() << std::endl;
```
std::cout

suo commented:
Out of curiosity: what happens if you create a class type that doesn't initialize its state?
```
@torch.jit.script
class Foo:
    x: int
    def __init__(self):
        pass
```

jamesr66a commented:
@suo looks like this:

```
class Foo:
  def __init__(self: __torch__.Foo) -> None:
    return None
```

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit b6b05e34:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=b6b05e342377cb521f8dc8a26fc9dbc375001d92).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4329368) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90106968) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90106968))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
+ git reset --hard b6b05e342377cb521f8dc8a26fc9dbc375001d92 
HEAD is now at b6b05e3423 Update on "[JIT] Serialize attributes and types in ClassType serialization" 
+ git merge --allow-unrelated-histories --no-edit --no-ff d2f66083c5b64e64355eb0fc580cfa4230bd17d8 
Auto-merging torch/custom_class.h 
CONFLICT (content): Merge conflict in torch/custom_class.h 
Auto-merging torch/csrc/jit/script/python_sugared_value.cpp 
Auto-merging test/test_jit.py 
CONFLICT (content): Merge conflict in test/test_jit.py 
Auto-merging test/cpp/jit/test_custom_class.cpp 
CONFLICT (content): Merge conflict in test/cpp/jit/test_custom_class.cpp 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]

Commit:
Update on "[JIT] Serialize attributes and types in ClassType serialization"


This allows us to properly serialize torchbind classes without having to dump and explicit `self.capsule = ...` thing in the `__init__` method

Differential Revision: [D19544737](https://our.internmc.facebook.com/intern/diff/D19544737)

[ghstack-poisoned]
